File: lib/djv/App/ToolsActions.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the DJV project.

#include <djv/App/ToolsActions.h>

#include <djv/App/App.h>
#include <djv/Models/SettingsModel.h>
#include <djv/Models/ToolsModel.h>

#include <algorithm>

#include <ftk/Core/Format.h>

namespace djv
{
    namespace app
    {
        struct ToolsActions::Private
        {
            std::shared_ptr<ftk::ListObserver<std::string> > openObserver;
        };

        void ToolsActions::_init(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app)
        {
            IActions::_init(context, app, "Tools");
            FTK_P();

            auto appWeak = std::weak_ptr<App>(app);
            for (const auto& tool : app->getToolsModel()->getTools())
            {
                // Register the command.
                _addCheckCommand(
                    tool.name,
                    ftk::Format("Toggle the {0} tool.").arg(tool.name),
                    [appWeak, tool](const nlohmann::json& args)
                    {
                        const bool value = args.at("value").get<bool>();
                        if (auto app = appWeak.lock())
                        {
                            app->getToolsModel()->setToolOpen(tool.name, value);
                            if (value)
                            {
                                // Opening a tool while the panel is hidden
                                // would otherwise do nothing that can be
                                // seen, which reads as the button being
                                // broken.
                                auto window =
                                    app->getSettingsModel()->getWindow();
                                if (!window.tools)
                                {
                                    window.tools = true;
                                    app->getSettingsModel()->setWindow(window);
                                }
                            }
                        }
                    });

                // Create the action.
                _actions[tool.name] = ftk::Action::create(
                    tool.name,
                    tool.icon,
                    _checkCommand(tool.name));

                // Register the shortcut.
                _addShortcut(tool.name, tool.name, tool.shortcut);
            }

            _shortcutsUpdate(app->getSettingsModel()->getShortcuts());

            p.openObserver = ftk::ListObserver<std::string>::create(
                app->getToolsModel()->observeOpenTools(),
                [this](const std::vector<std::string>& value)
                {
                    for (auto i : _actions)
                    {
                        i.second->setChecked(
                            std::find(value.begin(), value.end(), i.first) !=
                            value.end());
                    }
                });
        }

        ToolsActions::ToolsActions() :
            _p(new Private)
        {}

        ToolsActions::~ToolsActions()
        {}

        std::shared_ptr<ToolsActions> ToolsActions::create(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app)
        {
            auto out = std::shared_ptr<ToolsActions>(new ToolsActions);
            out->_init(context, app);
            return out;
        }
    }
}

File: examples/python/ToolsActions.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright Contributors to the DJV project.

import opentimelineio as otio
import ftkPy as ftk
import tlRenderPy as tl
import djvPy as djv

import IActions
import Tools
import Util

import weakref

class Actions(IActions.IActions):
    """
    This class provides the tool actions.
    """
    def __init__(self, context, app):
        IActions.IActions.__init__(self, context, app, "Tools")

        self._app = weakref.ref(app)
        self.tools = [
            tool for tool in app.getToolsModel().tools
            if tool.name in Tools.FACTORY]
        for tool in self.tools:
            self._addCheckCommand(
                tool.name,
                "Toggle the {} tool.".format(tool.name),
                lambda value, captured = tool.name, \
                    f = Util.weak(self._toolCallback): f(captured, value))
            self.actions[tool.name] = ftk.Action(
                tool.name,
                tool.icon,
                checkedCallback = self._checkCommand(tool.name))
            self._addShortcut(tool.name, tool.name, tool.shortcut)

        self._shortcutsUpdate(self._settingsModel.shortcuts)

        selfWeak = weakref.ref(self)
        self._openObserver = ftk.StringListObserver(
            app.getToolsModel().observeOpenTools,
            lambda names: selfWeak()._openUpdate(names))

    def _toolCallback(self, name, checked):
        app = self._app()
        app.getToolsModel().setToolOpen(name, checked)
        if checked:
            # Opening a tool while the panel is hidden would otherwise do
            # nothing that can be seen, which reads as the button being
            # broken.
            settingsModel = app.getSettingsModel()
            window = settingsModel.window
            if not window.tools:
                window.tools = True
                settingsModel.window = window

    def _openUpdate(self, names):
        for tool in self.tools:
            self.actions[tool.name].checked = tool.name in names
